java/src/main/java/org/rocksdb/AbstractComparator.java

```diff
@@ -52,52 +58,115 @@ protected AbstractComparator(final ComparatorOptions copt) {
+   * @param bLen the length of b, may be smaller than the buffer {@code b}
+   *
+   * @return the result of the comparison
+   */
+  private int compareInternal(final ByteBuffer a, final int aLen,
+      final ByteBuffer b, final int bLen) {
+    if (aLen != -1) {
+      a.mark();
+      a.limit(aLen);
+    }
+    if (bLen != -1) {
+      b.mark();
+      b.limit(bLen);
+    }
+
+    final int c = compare(a, b);
+
+    if (aLen != -1) {
+      a.reset();
+    }
+    if (bLen != -1) {
+      b.reset();
+    }
+
+    return c;
+  }
 
   /**
    * <p>Used to reduce the space requirements
    * for internal data structures like index blocks.</p>
    *
-   * <p>If start &lt; limit, you may return a new start which is a
+   * <p>If start &lt; limit, you may modify start which is a
    * shorter string in [start, limit).</p>
    *
-   * <p>Simple comparator implementations may return null if they
-   * wish to use start unchanged. i.e., an implementation of
-   * this method that does nothing is correct.</p>
+   * If you modify start, it is expected that you set the byte buffer so that
+   * a subsequent read of start.remaining() bytes from start.position()
+   * to start.limit() will obtain the new start value.
    *
-   * @param start String
-   * @param limit of type T
+   * <p>Simple comparator implementations may return with start unchanged.
+   * i.e., an implementation of this method that does nothing is correct.</p>
    *
-   * @return a shorter start, or null
+   * @param start the start
+   * @param limit the limit
    */
-  public String findShortestSeparator(final String start, final T limit) {
-      return null;
+  public void findShortestSeparator(final ByteBuffer start,
+      final ByteBuffer limit) {
+    // no-op
+  }
+
+  private int findShortestSeparatorInternal(
+      final ByteBuffer start, final int startLen,
+      final ByteBuffer limit, final int limitLen) {
+    if (startLen != -1) {
+      start.limit(startLen);
+    }
+    if (limitLen != -1) {
+      limit.limit(limitLen);
+    }
+    findShortestSeparator(start, limit);
+    return start.remaining();
   }
 
   /**
    * <p>Used to reduce the space requirements
    * for internal data structures like index blocks.</p>
    *
-   * <p>You may return a new short key (key1) where
+   * <p>You may change key to a shorter key (key1) where
    * key1 &ge; key.</p>
    *
-   * <p>Simple comparator implementations may return null if they
-   * wish to leave the key unchanged. i.e., an implementation of
+   * <p>Simple comparator implementations may return the key unchanged.
+   * i.e., an implementation of
    * this method that does nothing is correct.</p>
    *
-   * @param key String
-   *
-   * @return a shorter key, or null
+   * @param key the key
    */
-  public String findShortSuccessor(final String key) {
-      return null;
+  public void findShortSuccessor(final ByteBuffer key) {
+    // no-op
   }
+
+  private int findShortSuccessorInternal(
```

Comment:
Seems like these native-called private functions distract from the public API. Can they go in a package-protected class?

Reply:
I will investigate moving these into a internal helper class with static methods as discussed.